backend/oasst_backend/tree_manager.py

```diff
@@ -0,0 +1,760 @@
+import random
+from enum import Enum
+from typing import Optional, Tuple
+from uuid import UUID
+
+import numpy as np
+import pydantic
+from loguru import logger
+from oasst_backend.api.v1.utils import prepare_conversation, prepare_conversation_message_list
+from oasst_backend.config import settings
+from oasst_backend.models import Message, MessageReaction, MessageTreeState, TextLabels, message_tree_state
+from oasst_backend.prompt_repository import PromptRepository
+from oasst_backend.utils.hugging_face import HfEmbeddingModel, HfUrl, HuggingFaceAPI
+from oasst_shared.exceptions.oasst_api_error import OasstError, OasstErrorCode
+from oasst_shared.schemas import protocol as protocol_schema
+from sqlalchemy.sql import text
+from sqlmodel import Session, func
+
+
+class TreeManagerConfiguration(pydantic.BaseModel):
+    """Configuration class for the TreeManager"""
+
+    max_active_trees: int = 10
+    """Maximum number of concurrently active trees in the database.
+    No new initial prompt tasks will be handed out to users if this
+    number is reached."""
+
+    max_tree_depth: int = 6
+    """Maximum depth of message tree."""
+
+    max_children_count: int = 5
+    """Maximum number of reply messages per tree node."""
+
+    goal_tree_size: int = 15
+    """Total number of messages to gather per tree"""
+
+    num_reviews_initial_prompt: int = 3
+    """Number of peer review checks to collect in INITIAL_PROMPT_REVIEW state."""
+
+    num_reviews_reply: int = 3
+    """Number of peer review checks to collect per reply (other than initial_prompt)"""
+
+    acceptance_threshold_initial_prompt: float = 0.6
+    """Threshold for accepting an initial prompt"""
+
+    acceptance_threshold_reply: float = 0.6
+    """Threshold for accepting a reply"""
+
+    num_required_rankings: int = 3
+    """Number of rankings in which the message participated."""
+
+    mandatory_labels_initial_prompt: Optional[list[protocol_schema.TextLabel]] = [
```

Comment:
I feel we shouldn't ask "helpful" as mandatory, at least not for initial prompt and prompter reply, because it doesn't really apply to their situation. Maybe we also need to re-think the helpful label, maybe calling it "appropriate" or so, meaning that a prompt is really a prompt, an assistant response is of (helpful) assistance, etc.

Reply:
thinking a bit more, I think "appropriate" as I formulate it here would just be the opposite of "spam". the question is, if we know something is "not spam", in what way is that not enough, to the point where we'd need to collect another quality-gate label?

Reply:
ok, I removet `helpful` from all and also simplified the

```
def _calculate_acceptance(self, labels: list[TextLabels]):
        # calculate acceptance based on spam label
        return np.mean([1 - l.labels[protocol_schema.TextLabel.spam] for l in labels])
```